ws/src/frida_language_processing/scripts/item_categorization.py

```diff
@@ -1,152 +1,98 @@
-            api_key=os.getenv("OPENAI_API_KEY")
-        )
+        self._node = rclpy.create_node("item_categorization")
+        #services
+        #self._node.create_service(ItemsCategory,ITEMS_EXTRACT_TOPIC, self.get_local_embedding)
+        self._node.create_service(ItemsCategory, OBJECT_EXTRACT_TOPIC , self.extract_item)
+        #load the model
         self.model = SentenceTransformer("all-MiniLM-L12-v2")
         current_dir = os.path.dirname(os.path.abspath(__file__))
+        #Load the embeddings already calculated/ embeddings database
         with open(os.path.join(current_dir, "assets", "embeddings_items.pkl"), "rb") as f:
             self.items_dataframe = pickle.load(f)
 
         with open(os.path.join(current_dir, "assets", "embeddings_categories.pkl"), "rb") as f:
             self.category_embeddings = pickle.load(f)
-
-        rospy.spin()
+        
+        with open(os.path.join(current_dir, "assets", "embeddings_locations.pkl"), "rb") as f:
+            self.locations_embeddings = pickle.load(f)
+        
+        with open(os.path.join(current_dir, "assets", "embeddings_actions.pkl"), "rb") as f:
+            self.actions_embeddings = pickle.load(f)
+            #spin the node
+        rclpy.spin(self._node)
 
     def extract_item(self, req):
-        """Service to return the most representative category of the items given"""
-        target_item = req.items[0]
-        detected_items = req.items[1:]
-
-        item_embedding = self.get_embedding(target_item)
-        possible_items = self.get_similar_items(
-            item_embedding, self.items_dataframe)
-
-        response = ItemsCategoryResponse()
-
-        for i in range(len(possible_items)):
-            if possible_items.iloc[i]['name'] in detected_items:
-                response.category = possible_items.iloc[i]['name']
-                break
-
-        return response
-
-    def extract_category(self, req):
-        """Service to return the most representative category of the items given"""
-        items = req.items
-        categories = []
-        for item in items:
-            item_embedding = self.get_embedding(item)
-            item_category = self.get_category_similarity(
-                item_embedding, self.items_dataframe)
-            categories.append(item_category)
-
-        representative_category = max(set(categories), key=categories.count)
-        response = ItemsCategoryResponse()
-        response.category = representative_category
+        """Service to return the best match of the items given"""
+        start_time = time.time()
+        response = ItemsCategory.Response()
+        i = 0
+        for item in req.items:
+            item_embedding = [self.get_local_embedding(item)]
+            possible_items = self.get_similar_items(item_embedding[0], self.items_dataframe)
+            response.matches[i] = possible_items.iloc[0]['name']
+            i = 0 + 1
+        response.gpu_usage = self.get_gpu_usage()
```

Comment:
Maybe we can avoid computing details of gpu usage and response.time as they add overhead

Reply:
Also, gpu metrics could be affected if it's not the only process using gpu